<html>
  <head>
    <title>Futility</title>
    <link rel="shortcut icon" href="/favicon.ico" image="image/x-icon" />
  </head>
  <body>
    <h1>Futility</h1>
    <h2>What is it?</h2>
    <p>Welcome to fUtility (a utility for fitness). fUtility is an online tool for Tn-Seq data analysis and visualization. It has been designed to work with the phage-based platform for next gen sequencing of high-density transposon libraries described in <a href="http://www.biomedcentral.com/1471-2164/16/252">Santiago et al. 2015. BMC Genomics</a>. We have used this platform to identify genes conferring an increase or decrease in fitness in the presence of a set of antibiotics (both clinically-used antibiotics and probes). This dataset allows us to nominate mechanisms of action for compounds with an unknown target and to predict functions for hypothetical genes.</p>
    <h2>How do I use it?</h2>
    <p>There are multiple tools available. We have a fitness comparison table and a genome viewer. To get started, click the "Data Repository" tab and select a gene mapping.</p>
    <h3>Fitness Comparison Table</h3>
    <p>This table, available under the "Fitness" tab, can be used to compare the fitness of various treatments for a given gene. Simply select the gene of interest from the dropdown. Data can be sorted by clicking any of the column headers. To view the raw data for a gene, click on the name of the compound of interest, which will take you to the genome viewer.</p>
    <h3>Genome Viewer</h3>
    <p>Available under the "Raw Viewer" tab, this can be used to compare raw data between treatments. To begin, select a treatment from the dropdown and click "Add treatment." Raw data is colored by height to identify possible outliers (closer to red). You can zoom in and out of the raw data graphs by scrolling, and simply type in your gene of interest to move to another location in the genome.</p>
    <h2>Can I use my own data?</h2>
    <p>Yes! We can import custom genome labels (TODO describe format) and experiment data in .igv format. Once experiment data has been imported, it will be displayed in the fitness table and is available to be selected in the genome viewer. We default to also supplying data from our remote dataset, though you can turn this off by unchecking "Include Remote Data."</p>
    <h2>I have questions or comments</h2>
    <p>fUtility was designed by Suzanne Walkerâ€™s lab at Harvard Medical School, and was implemented by SolvedScience.</p>
    <p>For questions/comments on experimental protocols, please email suzanne_walker at <a href="http://hms.harvard.edu">hms.harvard.edu</a>.</p>
    <p>For questions/comments on the webtool, please contact SolvedScience (website and email)</p>
    <h2>Technical Details</h2>
    <p>TODO: technologies used, remote data set directory structure, formats used internally, tool for ingesting new data into remote data set.</p>
    <h3>Viewer Format</h3>
    <p>Thew viewer operates on a simple input that is derived directly from the igv file - one for each compound. It is a list of "read positions", which specify a start, end, and a number of reads. These are essentially the coordinates for building a histogram. We apply some minor optimizations here - we don't include any points where there were 0 reads, as we don't have to draw those, and we coalesce any neighboring, identical rectangles into a single rectangle of longer length. This format is generated in-browser when importing local data, and by a Ruby script when importing data into the curated set.</p>
    <h3>Fitness Table Format</h3>
    <p><strike>The fitness table is an interesting case because it operates on a slice across all compounds. When operating on local datasets, we take as input the gene and iterate over each compound, returning the properties at that gene. However, this would be very slow when accessing the curated set.</strike> When importing new data to the curated set, we create (if it doesn't already exist) a file for each gene. For each new compound, we iterate over the genes and append a row to the appropriate gene file. Then, when we want to load a gene from the remote data set, we just download the gene's file.</p>
  </body>
</html>
